server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,688 @@
+            .collect::<Result<_, ConversionToResponseError>>()?;
+    }
+
+    Ok(out)
+}
+
+struct IdempotencyKey(String);
+
+impl IdempotencyKey {
+    fn new(auth_token: &str, key: &str, url: &str) -> IdempotencyKey {
+        let mut hasher = Blake2b512::new();
+
+        hasher.update(auth_token);
+        hasher.update(":");
+        hasher.update(key);
+        hasher.update(":");
+        hasher.update(url);
+
+        let res = hasher.finalize();
+        IdempotencyKey(base64::encode(&res))
+    }
+}
+
+impl CacheKey for IdempotencyKey {
+    const PREFIX_CACHE: &'static str = "SVIX_IDEMPOTENCY_CACHE";
+}
+
+impl AsRef<str> for IdempotencyKey {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+async fn resolve_service<S>(
+    mut service: S,
+    req: Request<Body>,
+) -> Result<Response<BoxBody>, Infallible>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    service.call(req).await.map(IntoResponse::into_response)
+}
+
+/// The idempotency middleware itself -- used via the [`Router::layer`] method
+#[derive(Clone)]
+pub struct IdempotencyService<S: Clone> {
+    pub redis: Option<RedisCache>,
+    pub service: S,
+}
+
+impl<S> Service<Request<Body>> for IdempotencyService<S>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    type Response = Response<BoxBody>;
+    type Error = Infallible;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let mut service = self.service.clone();
+        let redis = self.redis.clone();
+
+        if let Some(redis) = redis {
+            Box::pin(async move {
+                let (parts, body) = req.into_parts();
+
+                // If not a POST request, simply resolve the service as usual
+                if parts.method != http::Method::POST {
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                }
+
+                // Retreive `IdempotencyKey` from header and URL parts, but returning the service
+                // normally in the event a key could not be created.
+                let key = if let Some(key) = get_key(&parts) {
+                    key
+                } else {
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                };
+
+                // Set the [`SerializedResponse::Start`] lock if the key does not exist in the cache
+                // returning whether the value was set
+                let set = if let Ok(set) = redis
```

Comment:
It took me multiple readings of this code and the function to understand that this variable means `lock_acquired`. Please rename it as such.